### DISCUSSION TITLE
How to obtain predictions from `LightningCLI` using the `predict` subcommand?

### DISCUSSION THREAD

**Discussion by christeefy** [Lightning-AI/pytorch-lightning#10509] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
First of all, I like to say it's very refreshing to see the changes to the library since I last used it in v1.1! Kudos to the team for all the great work 👏 

---

`pytorch-lightning` version: 1.5.0

## LightningCLI

I came across the `LightningCLI` class and found it to help reduce a lot of argument parsing boilerplate. 

My script looks like this:
```
# script.py
from pytorch_lightning.utilities.cli import LightningCLI

import ...

if __name__ == "__main__":
    LightningCLI(MyLightingModule, MyDataModule)
```

I was able to use it to `.fit` and `.test` my model, but I am not sure how to use `.predict`. 

When I pull up the help menu for this script, for the `predict` subcommand, I see that there's a flag to return predictions from the model:
```
>>> python -m script predict -h
⋮

Run inference on your data:
  --return_predictions {true,false,null}
                        Whether to return predictions. ``True`` by default except when an accelerator that
                        spawns processes is used (not supported). (type: Union[bool, null], default: null)
⋮
```
However, I cannot find any documentation on using `predict` with the `LightningCLI` or on how to obtain a handle to the predictions. 

Appreciate guidance on how to do so and thanks in advance! 😊

**Comment by mauvilsa** [comment]
Create a Callback class that implements [on_predict_epoch_end](https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#on-predict-epoch-end). This method receives as argument the predicted results, so there you can do whatever is needed. Then add this callback to the CLI either as non-configurable by using [trainer_defaults](https://pytorch-lightning.readthedocs.io/en/latest/api/pytorch_lightning.utilities.cli.html#pytorch_lightning.utilities.cli.LightningCLI.params.trainer_defaults) as `LightningCLI(MyLightingModule, MyDataModule, trainer_defaults={'callbacks': [MyCallback()]})` or make it a [configurable callback](https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_cli.html#configurable-callbacks).

**Comment by kaczmarj** [comment]
look like the suggested method as of may 2023 is to use `lightning.pytorch.callbacks.BasePredictionWriter` in conjunction with a `lightning.Trainer`.

Example from [the docs](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.callbacks.BasePredictionWriter.html#lightning.pytorch.callbacks.BasePredictionWriter
):

```python
import torch
from lightning.pytorch.callbacks import BasePredictionWriter

class CustomWriter(BasePredictionWriter):

    def __init__(self, output_dir, write_interval):
        super().__init__(write_interval)
        self.output_dir = output_dir

    def write_on_batch_end(
        self, trainer, pl_module', prediction, batch_indices, batch, batch_idx, dataloader_idx
    ):
        torch.save(prediction, os.path.join(self.output_dir, dataloader_idx, f"{batch_idx}.pt"))

    def write_on_epoch_end(self, trainer, pl_module, predictions, batch_indices):
        torch.save(predictions, os.path.join(self.output_dir, "predictions.pt"))


pred_writer = CustomWriter(output_dir="pred_path", write_interval="epoch")
trainer = Trainer(callbacks=[pred_writer])
model = BoringModel()
trainer.predict(model, return_predictions=False)
```

one would replace the last three lines with a `LightningCLI` object and `trainer_defaults={'callbacks': [CustomWriter(output_dir)]}`
